### ISSUE TITLE
Allow the Insertion of a Step into the Building of a Distributable

### ISSUE THREAD

**Issue by ArsThaumaturgis** [panda3d/panda3d#1497]
## Description
In short, I feel that it may be useful to be able to specify a custom step in the building process, taking place between the population of the "build" directory and the packaging of its contents into the "dist" directory.

I imagine this as perhaps an optional callback (maybe specified in "setup.py"?) to a Python-script provided by the developer.

## Use Case
There may be times in which a project wants for additional processing in the making of a distributable, and specifically done after the generation of its final file-structure in the "build" directory, but before everything is packed up into the "dist" directory. This might be to convert files, to adjust file structure, or otherwise to alter things to suit the shift from the environment of "development" to the environment of "distribution".

For example, in a project of my own in which I want to pack certain models into a multifile for distribution. These are, naturally, originally in "egg" format, and I would prefer of course that in distribution they be instead in the optimised "bam" format.

The build system already has the functionality to perform this conversion--but not if the files are packed into a multifile. Naturally, one answer to this is to convert them myself before packing them into the multifile.

But it would be handy, I feel, to instead simply place the packing step between the "build" step--which would convert the files as a matter of course--and the "dist" step.

**Comment by Moguri** [comment]
I agree that more control/flexibility in the build pipeline would be nice, which is partly why I made and use [pman]( ). As a work-around, you can define a custom `build_apps` that runs your extra logic after Panda's `build_apps`. Something like:


```python
from setuptools import setup

from direct.dist.commands import build_apps

class BuildApps(build_apps):
    def run(self):
        # Run regular build_apps
        build_apps.run(self)

        # Add extra code here

setup(
    ...
    cmdclass={
        'build_apps': BuildApps,
    },
    ...
)
```

**Comment by rdb** [comment]
I'd be in favour of a more pluggable, flexible, modular build infrastructure.  I think some design effort would need to go into this.
